File: android/app/src/test/kotlin/com/neovibe/neovibe/RecognitionVectorsTest.kt
package com.neovibe.neovibe

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.neovibe.neovibe.ble.AdvertSchedule
import com.neovibe.neovibe.ble.BleConstants
import com.neovibe.neovibe.ble.RecognitionTable
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import java.io.File
import java.util.Base64

/**
 * **Le cote KOTLIN du point de contact avec le Dart.**
 *
 * ## 🔴 Le trou que ce test bouche — releve le 2026-08-29
 *
 * Le plan d'emission et la table de reconnaissance sont **ecrits en Dart et lus
 * ici**. `AdvertScheduleTest` et `SightingBookTest` eprouvent ce fichier-ci avec
 * **leurs propres fixtures** ; `advert_plan_test.dart` eprouve l'autre cote avec
 * les siennes. Rien ne verifiait qu'ils rangent les octets de la meme facon.
 *
 * ⚠️ **Une transposition serait parfaitement silencieuse.** Si on lisait la
 * table « ami par ami » la ou le Dart l'ecrit « creneau par creneau », tout
 * compilerait, les tests des deux cotes resteraient verts, et le seul symptome
 * serait : *l'appareil est entendu dix fois par seconde et reconnu zero fois*.
 *
 * ⚠️ **Ces vecteurs ne se regenerent pas pour faire passer un test.** Un echec
 * ici veut dire « le format a change cote Dart » — et la question suivante est
 * « ce fichier a-t-il change avec ? ». Regenerer sans y repondre laisse les deux
 * implementations fausses **ensemble**, ce qui est pire qu'une seule fausse :
 * plus rien ne les departage.
 *
 * Production des vecteurs :
 * `NEOVIBE_REGEN=1 flutter test test/recognition_vectors_test.dart`
 */
class RecognitionVectorsTest {

    private val manifeste: JsonObject by lazy {
        // ⚠️ Chemin depuis `android/app` — la racine d'execution des tests JVM.
        val f = File("src/test/resources/recognition-vectors/manifest.json")
        check(f.exists()) {
            "Vecteurs absents (" + f.absolutePath + "). Les generer avec : " +
                "NEOVIBE_REGEN=1 flutter test test/recognition_vectors_test.dart"
        }
        JsonParser.parseString(f.readText()).asJsonObject
    }

    private val slotMillis: Long get() = manifeste["slotMillis"].asLong
    private val tokenLength: Int get() = manifeste["tokenLength"].asInt

    private fun octets(base64: String): ByteArray = Base64.getDecoder().decode(base64)

    private fun hex(bytes: ByteArray): String =
        bytes.joinToString("") { "%02x".format(it) }

    private fun deHex(s: String): ByteArray =
        ByteArray(s.length / 2) { ((s[it * 2].digitToInt(16) shl 4) or s[it * 2 + 1].digitToInt(16)).toByte() }

    private fun table(): RecognitionTable {
        val t = manifeste["table"].asJsonObject
        return RecognitionTable(
            tableId = t["tableId"].asInt,
            fromSlot = t["fromSlot"].asLong,
            slotMillis = slotMillis,
            slotCount = t["slotCount"].asInt,
            perSlot = t["perSlot"].asInt,
            tokens = octets(t["tokens"].asString),
            tokenLength = tokenLength,
        )
    }

    private fun plan(): AdvertSchedule {
        val p = manifeste["plan"].asJsonObject
        return AdvertSchedule(
            fromSlot = p["fromSlot"].asLong,
            slotMillis = slotMillis,
            slotCount = p["slotCount"].asInt,
            perSlot = p["perSlot"].asInt,
            tokens = octets(p["tokens"].asString),
            tokenLength = tokenLength,
            types = octets(p["types"].asString),
        )
    }

    // ------------------------------------------------------------------
    // La reconnaissance
    // ------------------------------------------------------------------

    @Test
    fun `chaque sonde du Dart trouve ce que le Dart annonce`() {
        val table = table()
        val sondes = manifeste["sondes"].asJsonArray
        check(sondes.size() >= 5) { "vecteurs trop pauvres pour prouver quoi que ce soit" }

        for (element in sondes) {
            val sonde = element.asJsonObject
            val pourquoi = sonde["pourquoi"].asString
            // Milieu du creneau : on eprouve la regle, pas un effet de bord.
            val instant = sonde["auCreneau"].asLong * slotMillis + slotMillis / 2
            val rang = table.match(deHex(sonde["jeton"].asString), instant)

            if (sonde["rangAttendu"].isJsonNull) {
                assertNull(pourquoi, rang)
            } else {
                assertEquals(pourquoi, sonde["rangAttendu"].asInt, rang)
            }
        }
    }

    @Test
    fun `la taille de jeton est celle que l'annonce transporte`() {
        // ⚠️ Une divergence ici rendrait TOUTE reconnaissance impossible, sur
        // tous les appareils, sans lever la moindre erreur. `ADVERT_PAYLOAD_SIZE`
        // = 2 octets de magie + 1 de version + 1 de type + le jeton.
        assertEquals(
            "le jeton du Dart ne tient pas dans l'annonce du natif",
            BleConstants.ADVERT_PAYLOAD_SIZE - 4,
            tokenLength,
        )
    }

    // ------------------------------------------------------------------
    // Le plan d'emission
    // ------------------------------------------------------------------

    @Test
    fun `le plan rend les jetons du Dart, dans l'ordre du Dart`() {
        val plan = plan()
        for (element in manifeste["sondesPlan"].asJsonArray) {
            val sonde = element.asJsonObject
            val creneau = sonde["auCreneau"].asLong
            val instant = creneau * slotMillis + slotMillis / 2

            val duCreneau = plan.tokensAt(instant, avecPublic = true)
            checkNotNull(duCreneau) { "le plan ne couvre pas le creneau " + creneau }

            val attendus = sonde["jetons"].asJsonArray.map { it.asString }
            assertEquals(
                "les jetons du creneau " + creneau,
                attendus,
                duCreneau.first.map { hex(it) },
            )

            val typesAttendus = sonde["types"].asJsonArray.map { it.asInt.toByte() }
            assertEquals(
                "les TYPES du creneau " + creneau + " — un jeton d'ami pris pour " +
                    "un identifiant public rend l'appareil inconnaissable",
                typesAttendus,
                duCreneau.second.toList(),
            )
        }
    }

    @Test
    fun `sans le droit de crier l'identifiant public, il ne reste que les amis`() {
        // ⚠️ L'homme mort de l'identifiant public (2026-08-28) : le Dart mort,
        // le jeton public ne vaut plus rien et doit cesser de partir — les
        // jetons d'AMIS, eux, continuent douze heures.
        val plan = plan()
        for (element in manifeste["sondesPlan"].asJsonArray) {
            val sonde = element.asJsonObject
            val instant = sonde["auCreneau"].asLong * slotMillis + slotMillis / 2
            val duCreneau = plan.tokensAt(instant, avecPublic = false)
            checkNotNull(duCreneau)

            assertEquals(
                "seuls les jetons d'amis restent",
                sonde["sansPublic"].asJsonArray.map { it.asString },
                duCreneau.first.map { hex(it) },
            )
            assertEquals(
                "aucun jeton public ne doit subsister",
                emptyList<Byte>(),
                duCreneau.second.toList().filter { it == BleConstants.TYPE_PUBLIC },
            )
        }
    }
}

File: test/recognition_vectors_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:neovibe/features/proximity/net/advert_plan.dart';
import 'package:neovibe/features/proximity/proximity_identity.dart';

/// **Le point de contact entre le Dart qui ÉCRIT et le Kotlin qui LIT.**
///
/// ## 🔴 Le trou que ces vecteurs bouchent — relevé le 2026-08-29
///
/// Le plan d'émission et la table de reconnaissance sont écrits en Dart et lus
/// en Kotlin. Les deux côtés avaient leurs tests — `advert_plan_test.dart` ici,
/// `AdvertScheduleTest` et `SightingBookTest` là-bas — **mais chacun avec ses
/// propres fixtures**. Rien ne vérifiait qu'ils rangent les octets de la même
/// façon.
///
/// ⚠️ **Une transposition y serait parfaitement silencieuse.** Si Kotlin lisait
/// la table « ami par ami » là où Dart l'écrit « créneau par créneau », tout
/// compilerait, les tests des deux côtés resteraient verts, et le seul symptôme
/// serait : *l'appareil est entendu dix fois par seconde et reconnu zéro fois*.
/// C'est le symptôme exact qu'on a passé la journée du 2026-08-29 à
/// diagnostiquer — il avait cette fois une autre cause, mais rien n'aurait
/// permis de distinguer les deux.
///
/// Même dispositif que pour le format scellé (`tool/gen_seal_vectors.dart` +
/// `SealedChunkReaderTest`), et pour la même raison : **deux implémentations
/// d'un même accord ont besoin d'un point de contact.**
///
/// ## ⚠️ Pourquoi c'est un TEST et pas un `tool/gen_*.dart`
///
/// `ProximityIdentity` importe `flutter_secure_storage` et `path_provider` :
/// `dart run` ne sait pas le compiler. Il faut le moteur Flutter, donc
/// `flutter test`.
///
/// **Et c'est mieux ainsi** : un générateur ne vérifie rien après le jour où on
/// l'a lancé. Ici le côté Dart est **contrôlé à chaque exécution de la suite** —
/// si quelqu'un change l'ordre d'aplatissement ou la formule d'un jeton, ce
/// test tombe *avant* que l'appareil ne devienne muet.
///
/// ## Régénérer — délibérément, jamais « pour faire passer le test »
///
/// ```
/// NEOVIBE_REGEN=1 flutter test test/recognition_vectors_test.dart
/// ```
///
/// ⚠️ **Régénérer pour faire taire un échec annule tout le dispositif** : les
/// deux implémentations resteraient fausses ensemble. Un échec ici veut dire
/// *« le format a changé »* — et la question suivante est *« le Kotlin a-t-il
/// changé avec ? »*.
const _fichier =
    'android/app/src/test/resources/recognition-vectors/manifest.json';

/// Un créneau fixe : des vecteurs qui changent à chaque exécution ne prouvent
/// rien, et leur diff devient illisible.
const _depart = 1986600;

/// ⚠️ **Trois créneaux au moins, et DEUX amis au moins.** Avec un seul de l'un
/// ou de l'autre, une table transposée donnerait exactement les mêmes octets :
/// le vecteur ne prouverait rien.
const _creneaux = 3;

const _moi = 'u-charles';

final _amis = <String, Uint8List>{
  'u-mimi': Uint8List.fromList(List.generate(32, (i) => i)),
  'u-lea': Uint8List.fromList(List.generate(32, (i) => 255 - i)),
};

final _grainePing = Uint8List.fromList(
  List.generate(32, (i) => (i * 7) & 0xFF),
);

Future<Map<String, dynamic>> _construire() async {
  const planificateur = AdvertPlanner();

  final table = await planificateur.nativeTable(
    secrets: _amis,
    fromSlot: _depart,
    slots: _creneaux,
    tableId: 42,
  );

  final plan = await planificateur.plan(
    secrets: _amis,
    fromSlot: _depart,
    slots: _creneaux,
    meUserId: _moi,
    pingSeed: _grainePing,
  );

  // ⚠️ **Le même aplatissement que `ProximitySupervisor._deposePlan`.** S'il
  // divergeait, ces vecteurs décriraient un plan que personne ne dépose — un
  // instrument qui mesure autre chose que ce qui tourne.
  final planPlat = Uint8List(
    plan.tokens.length * ProximityIdentity.tokenLength,
  );
  final types = Uint8List(plan.tokens.length);
  for (var i = 0; i < plan.tokens.length; i++) {
    planPlat.setRange(
      i * ProximityIdentity.tokenLength,
      (i + 1) * ProximityIdentity.tokenLength,
      plan.tokens[i].bytes,
    );
    types[i] = plan.tokens[i].audience == null ? 1 : 2;
  }

  String hex(Uint8List b) => ProximityIdentity.hex(b);
  Future<String> jetonDe(String ami, int creneau) async => hex(
    await ProximityIdentity.pairToken(_amis[ami]!, creneau, emitter: ami),
  );

  final sondes = <Map<String, dynamic>>[
    {
      'jeton': await jetonDe('u-mimi', _depart),
      'auCreneau': _depart,
      'rangAttendu': 0,
      'pourquoi': 'le premier ami, a son propre creneau',
    },
    {
      // 🔴 **La sonde qui attrape une transposition.** Le rang 1 au troisieme
      // creneau n'est au bon endroit que si les deux cotes rangent creneau par
      // creneau. Transposee, la table y trouverait quelqu'un d'autre.
      'jeton': await jetonDe('u-lea', _depart + 2),
      'auCreneau': _depart + 2,
      'rangAttendu': 1,
      'pourquoi': 'le SECOND ami au TROISIEME creneau — la transposition',
    },
    {
      'jeton': await jetonDe('u-mimi', _depart + 1),
      'auCreneau': _depart,
      'rangAttendu': 0,
      'pourquoi': 'un creneau d avance : la tolerance d horloge l accepte',
    },
    {
      'jeton': await jetonDe('u-mimi', _depart),
      'auCreneau': _depart + 2,
      'rangAttendu': null,
      'pourquoi': 'deux creneaux de retard : rejoue, donc refuse',
    },
    {
      'jeton': hex(
        await ProximityIdentity.pairToken(
          Uint8List.fromList(List.filled(32, 9)),
          _depart,
          emitter: 'u-inconnu',
        ),
      ),
      'auCreneau': _depart,
      'rangAttendu': null,
      'pourquoi': 'le jeton d une autre paire',
    },
    {
      // ⚠️ Deux formats, deux chemins (consigne de Jay, 2026-08-25) : le jeton
      // public n'entre jamais dans la table de reconnaissance.
      'jeton': hex(await ProximityIdentity.publicPingId(_grainePing, _depart)),
      'auCreneau': _depart,
      'rangAttendu': null,
      'pourquoi': 'l identifiant PUBLIC n est reconnu par personne',
    },
    {
      // ⚠️ **Le SENS du jeton.** Ce que moi j'emets vers mimi n'est pas ce que
      // j'attends d'elle. Une inversion ne leve aucune erreur : elle rend
      // simplement les deux appareils sourds l'un a l'autre.
      'jeton': hex(
        await ProximityIdentity.pairToken(
          _amis['u-mimi']!,
          _depart,
          emitter: _moi,
        ),
      ),
      'auCreneau': _depart,
      'rangAttendu': null,
      'pourquoi': 'mon PROPRE jeton vers mimi : je ne dois pas le reconnaitre',
    },
  ];

  final sondesPlan = <Map<String, dynamic>>[
    for (var c = _depart; c < _depart + _creneaux; c++)
      {
        'auCreneau': c,
        'jetons': [for (final t in plan.forSlot(c)) hex(t.bytes)],
        'types': [for (final t in plan.forSlot(c)) t.audience == null ? 1 : 2],
        'sansPublic': [
          for (final t in plan.forSlot(c))
            if (t.audience != null) hex(t.bytes),
        ],
      },
  ];

  return {
    'slotMillis': ProximityIdentity.slotDuration.inMilliseconds,
    'tokenLength': ProximityIdentity.tokenLength,
    'table': {
      'tableId': table.tableId,
      'fromSlot': table.fromSlot,
      'slotCount': table.slotCount,
      'perSlot': table.perSlot,
      'ordre': table.order,
      'tokens': base64.encode(table.tokens),
    },
    'plan': {
      'fromSlot': plan.fromSlot,
      'slotCount': _creneaux,
      'perSlot': plan.forSlot(_depart).length,
      'tokens': base64.encode(planPlat),
      'types': base64.encode(types),
    },
    'sondes': sondes,
    'sondesPlan': sondesPlan,
  };
}

void main() {
  test('les vecteurs croisés décrivent le format que le Dart produit', () async {
    final attendu =
        '${const JsonEncoder.withIndent('  ').convert(await _construire())}\n';
    final fichier = File(_fichier);

    if (Platform.environment['NEOVIBE_REGEN'] == '1') {
      fichier.parent.createSync(recursive: true);
      fichier.writeAsStringSync(attendu);
      // ignore: avoid_print
      print(
        'VECTEURS RÉGÉNÉRÉS dans $_fichier.\n'
        '⚠️ Vérifie que le Kotlin a changé avec le format, sinon les deux '
        'implémentations sont fausses ensemble.',
      );
      return;
    }

    expect(
      fichier.existsSync(),
      isTrue,
      reason:
          'Vecteurs absents. Générer avec : '
          'NEOVIBE_REGEN=1 flutter test test/recognition_vectors_test.dart',
    );
    expect(
      fichier.readAsStringSync().replaceAll('\r\n', '\n'),
      attendu,
      reason:
          'Le format d\'émission ou de reconnaissance a changé côté Dart. '
          'Le Kotlin a-t-il changé avec ? Régénérer sans vérifier laisserait '
          'les deux implémentations fausses ensemble.',
    );
  });
}
